src/wrapper_generators.jl

```diff
@@ -12,7 +17,21 @@ macro generate_wrapper_header(src_name)
             else
                 # We explicitly use `macrocall` here so that we can manually pass the `__source__`
                 # argument, to avoid `@artifact_str` trying to lookup `Artifacts.toml` here.
-                return $(Expr(:macrocall, Symbol("@artifact_str"), __source__, src_name))
+                return $(artifact_expr)
+            end
+        end
+        function find_or_instantiate_artifact_dir()
+            @static if VERSION < v"1.6"
+                return find_artifact_dir()
+            else
+                try
+                    return find_artifact_dir()
+                catch
+                    # 1.6 does not call `.pkg/select_artifacts.jl`
+                    # 1.7+ User might switch the tag, without calling `Pkg.instantiate()`
```

Comment:
@staticfloat :cry:

Reply:
`Pkg` should not be loaded by most JLLs, so the call to `ensure_artifact_installed()` will likely fail.  I'm kind of thinking if this is a concern for your package, you should be forced to use `LazyArtifacts` all the way, and then we don't worry about `Pkg.add()`-time installs; everything just happens dynamically at runtime.

Sorry, I didn't realize this platform tag was something that could change quickly like that; I thought this was more of a system property, e.g. like `libgfortran` version, which we expect to essentially never change, and so if it did, it was okay if things give strange errors.

Reply:
Hm currently all `JLL` packages load `Pkg`. I know that it was the intention to remove the reliance on it, but shrug. (https://github.com/JuliaBinaryWrappers/LAMMPS_jll.jl/blob/9847851f2f33f6e14572e76c42d183b1f9f4af3f/Project.toml#L7)

Yeah the earlier variant of the commit was called "And suddenly all Artifacts are Lazy". My primary concern here is the user experience. I would really like it to just work. Telling people to "restart" Julia is borderline okay. Having to call `Pkg.instantiate()` manually is a step that I don't want to do. 

The other problem is 1.6, there it simply will fail, since that facility of doesn't exist. So I thought of turning the problem into a blessing in disguise.

Reply:
> Hm currently all JLL packages load Pkg.

Do they https://github.com/JuliaPackaging/JLLWrappers.jl/blob/dd045e0576ddf834754100af4cc4829b236ae801/Project.toml#L6-L7? 🤔

Reply:
> ok, the JLL themselve load Pkg

Yeah was about to say. JLLWrappers.jl doesn't depend on Pkg, but the JLLs currently do.

Reply:
>Sorry, I didn't realize this platform tag was something that could change quickly like that; I thought this was more of a system property

Yeah for MPI, CUDA and a few other places it is a user choice to go tell us to use a different implementation. I would be amendable to removing this for 1.7 and then we need to solve this problem elsewhere. But I am not seeing how this will work on 1.6 without either fixing this or backporting other changes.

Reply:
> Hm currently all JLL packages load Pkg.

They _depend_ on `Pkg`, because there is no separate `Artifacts` module on Julia 1.5-, so [we must use `Pkg` to get access to `Artifacts`](https://github.com/JuliaPackaging/JLLWrappers.jl/blob/master/src/toplevel_generators.jl#L28-L34).  But on Julia 1.6+, we don't ever load `Pkg`, and that's a big reason we're able to load JLLs so fast.  If we load `Pkg`, we automatically incur a lot of startup time.

> But I am not seeing how this will work on 1.6 without either fixing this or backporting other changes.

We can use `LazyArtifacts` and force all MPI-related things to be downloaded at usage time, once the package actually calls `artifact"foo"`.

Reply:
I could delay the loading unttil the moment where we know that we need to instantiate the artifact?

Reply:
Especially for MPI lazy loading is problematic since we really don't want a 100 ranks hitting the file system at the same time